### PR TITLE
🧙‍♂️ Link based loading of instances

### PR DESCRIPTION
**Pull request by nikokaoja**
This shows what we can do now:

Screenshot 2024-12-03 at 21 24 33

### CODE REVIEW

File: cognite/neat/_store/_base.py [hunk 1]
```diff
@@ -191,25 +191,50 @@ def write(self, extractor: TripleExtractors) -> None:
...
-        """Read instances for given view from the graph store."""
-
-        if not self.rules:
+    def _read_via_rules_linkage(
+        self, class_neat_id: URIRef, property_link_pairs: dict[str, URIRef] | None
+    ) -> Iterable[tuple[str, dict[str | InstanceType, list[str]]]]:
+        if self.rules is None:
             warnings.warn("Rules not found in graph store!", stacklevel=2)
-            return None
+            return []
```
Comment (nikokaoja) [inline cognite/neat/_store/_base.py:199]: mypy complains, at one place I am forced to put [] , at other None

File: tests/tests_unit/graph/test_loaders/test_dms_loader.py [hunk 2]
```diff
@@ -15,16 +16,36 @@ def test_metadata_as_json_filed():
 
     importer = InferenceImporter.from_graph_store(store)
 
-    rules = ImporterPipeline.verify(importer)
-    store.add_rules(rules)
+    info_rules = ImporterPipeline.verify(importer)
+    dms_rules = InformationToDMS().transform(info_rules).rules
+
```
Comment (nikokaoja) [inline tests/tests_unit/graph/test_loaders/test_dms_loader.py:21]: This test simulates the process of updating data model on both ends but still being able to get what you are expecting

File: cognite/neat/_graph/loaders/_rdf2dms.py [hunk 3]
```diff
@@ -244,7 +284,7 @@ def _create_node(
         return dm.NodeApply(
             space=self.instance_space,
             external_id=identifier,
-            type=dm.DirectRelationReference(view_id.space, type_) if type_ is not None else None,
```
Comment (nikokaoja) [inline cognite/neat/_graph/loaders/_rdf2dms.py:247]: this was left-over from days when we only supported 1-1 between info and dms rules

File: cognite/neat/_graph/loaders/_rdf2dms.py [hunk 4]
```diff
@@ -114,17 +135,36 @@ def _load(self, stop_on_exception: bool = False) -> Iterable[dm.InstanceApply |
...
+                    reader = self.graph_store._read_via_rules_linkage(class_neat_id, property_link_pairs)
+                else:
+                    error_view = ResourceRetrievalError(view_id, "view", "View not linked to class")
+                    tracker.issue(error_view)
+                    if stop_on_exception:
+                        raise error_view
+                    yield error_view
+
+            else:
+                print("Loading instances without linked properties")
```
Comment (doctrino) [inline cognite/neat/_graph/loaders/_rdf2dms.py:155]:
```suggestion
```

**Review by doctrino — APPROVED** [review]

**Pull request merged**